timemachine/fe/rbfe.py

```diff
@@ -311,23 +326,29 @@ def estimate_free_energy_given_initial_states(initial_states, protocol, temperat
 
         if lamb_idx > 0:
 
-            # loop over bond, angle, torsion, nonbonded terms etc.
-            all_fwd_delta_us = []
-            all_rev_delta_us = []
+            ukln_by_term = []
 
+            # loop over bond, angle, torsion, nonbonded terms etc.
             for u_idx, (prev_U_fn, cur_U_fn) in enumerate(zip(prev_batch_U_fns, cur_batch_U_fns)):
-                fwd_delta_u = beta * (cur_U_fn(prev_frames, prev_boxes) - prev_U_fn(prev_frames, prev_boxes))
-                rev_delta_u = beta * (prev_U_fn(cur_frames, cur_boxes) - cur_U_fn(cur_frames, cur_boxes))
+                x_cur = (cur_frames, cur_boxes)
+                x_prev = (prev_frames, prev_boxes)
+                u_00 = beta * prev_U_fn(*x_prev)
+                u_01 = beta * prev_U_fn(*x_cur)
+                u_10 = beta * cur_U_fn(*x_prev)
+                u_11 = beta * cur_U_fn(*x_cur)
+                ukln_by_term.append([[u_00, u_01], [u_10, u_11]])
+
+                fwd_delta_u = u_10 - u_00
+                rev_delta_u = u_01 - u_11
                 df, df_err = pymbar.BAR(fwd_delta_u, rev_delta_u)
                 plot_axis = all_axes[lamb_idx - 1][u_idx]
                 plot_BAR(df, df_err, fwd_delta_u, rev_delta_u, U_names[u_idx], plot_axis)
-                all_fwd_delta_us.append(fwd_delta_u)
-                all_rev_delta_us.append(rev_delta_u)
 
             # sanity check - I don't think the dG calculation commutes with its components, so we have to re-estimate
             # the dG from the sum of the delta_us as opposed to simply summing the component dGs
-            total_fwd_delta_us = np.sum(all_fwd_delta_us, axis=0)
-            total_rev_delta_us = np.sum(all_rev_delta_us, axis=0)
+            ukln_by_term = np.array(ukln_by_term, dtype=np.float64)
+            total_fwd_delta_us = (ukln_by_term[:, 1, 0, :] - ukln_by_term[:, 0, 0, :]).sum(axis=0)
+            total_rev_delta_us = (ukln_by_term[:, 0, 1, :] - ukln_by_term[:, 1, 1, :]).sum(axis=0)
```

Comment:
I found this difficult to read at first, but from comparison to L335-339 it looks correct.

Suggestion: add a comment to `ukln_by_term` to name the 4 axes
```suggestion
            # (num components, energy fxn idx, sampled state idx, num samples)
            ukln_by_term = np.array(ukln_by_term, dtype=np.float64)
            total_fwd_delta_us = (ukln_by_term[:, 1, 0, :] - ukln_by_term[:, 0, 0, :]).sum(axis=0)
            total_rev_delta_us = (ukln_by_term[:, 0, 1, :] - ukln_by_term[:, 1, 1, :]).sum(axis=0)
```

Reply:
Thanks! Applied in [68229f1](https://github.com/proteneer/timemachine/pull/811/commits/68229f10602559de212a3a24cff6bf7bf78b331f). I also slightly prefer your usage of "component" instead of "term" and updated the naming in [2dd0942](https://github.com/proteneer/timemachine/pull/811/commits/2dd09428c53aac1fb88d0087c88f8a0e6b305663).